### PR TITLE
Fix MoA edge-case stability and add boundary regression tests

### LINKED ISSUE #53: 2026犀牛鸟开源人才专属】【低难度】MoA(mixture of Attention)：边界测试补全与垂类训练验证
在现有 tests/test_moa.py 基础上，系统补全 MoA 模块的边界测试与训练验证，提升测试覆盖率
测试补全（至少覆盖以下场景）：
NeckMoAFusion 在跨尺度输入尺寸不匹配（如 hi 为 15×15、lo 为 7×7，非严格 2× 下采样）时的前向稳定性与形状保持
MoABlock 的 temperature 退火到极小值（如 temperature < 1e-4）时，softmax 路由概率的数值稳定性（是否出现 NaN 或均匀分布）
_LocalAttnHead 与 _GlobalAttnHead 在 num_heads 不能被 dim 整除时的降级处理（_safe_groups 的边界）
C2fMoA 的 aux_loss 在多 MoABlock 嵌套时是否存在重复计数（类似 MoE 的 MOE_LOSS_REGISTRY 双计数问题）
缺陷修复：在补充测试过程中，若发现任何边界缺陷（如 IndexError、NaN 传播、形状不匹配），需一并定位并修复
覆盖率报告：提供 pytest --cov 前后的覆盖率对比，至少覆盖 ultralytics/nn/modules/moa/ 目录
垂类训练验证：在 VisDrone 或 SKU-110K 上，使用 YOLO-Master-v0.10-MoA-N 训练 50~100 epoch，验证 MoA 模块在真实数据集上的收敛性，记录 mAP50-95 与 loss 曲线，与同配置的 MoE 基线对比
提交 Pull Request，包含测试代码、修复代码、覆盖率报告、训练日志
https://github.com/Tencent/YOLO-Master
https://github.com/Tencent/YOLO-Master/blob/main/ultralytics/nn/modules/moa/moa.py
https://github.com/Tencent/YOLO-Master/blob/main/tests/test_moa.py
https://github.com/Tencent/YOLO-Master/blob/main/ultralytics/cfg/models/master/v0_10/det/yolo-master-moa-n.yaml
https://github.com/Tencent/YOLO-Master/blob/main/scripts/compare_moa_ablation.py
难度：低

---

**本issue为2026犀牛鸟开源人才培养活动专属issue，仅供已报名参与犀牛鸟活动的同学认领**
【认领时间】7月1日～7月31日（7月1日前认领视为无效❗️）
【认领方式】在本issue评论区回复“已认领本任务”，即视为认领成功
【活动报名】需提前完成犀牛鸟报名问卷，问卷将用于活动登记和奖励发放：https://wj.qq.com/s2/26888567/gh2q

### PR DESCRIPTION
**Pull request by Alanxtl**
fix: #53

## 概述

本 PR 主要补全 MoA 模块的边界测试，并修复在测试和 VisDrone 训练验证中发现的一些数值稳定性与形状对齐问题。

还包含4个实验结果 分别是 moe/moa 在 50epoch/100epoch 上的 visdrone 训练结果

实验环境：Ultralytics 8.3.240 🚀 Python-3.11.15 torch-2.12.1+cu130 CUDA:0 (NVIDIA GeForce RTX 3090, 24126MiB)

### 测试补全

在 `tests/test_moa.py` 中新增/补充了以下测试：

- `NeckMoAFusion` 跨尺度尺寸不严格匹配测试
  - 覆盖 `hi = 15x15`、`lo = 7x7` 的情况
  - 验证 forward 不报错
  - 验证输出 shape 与 `hi` 保持一致
  - 验证输出为有限值，并且 backward 梯度正常

- `MoABlock` 极小 temperature 路由稳定性测试
  - 覆盖 `temperature = 1e-6`
  - 验证 router logits / softmax 概率不会出现 `NaN` 或 `inf`
  - 验证概率和为 1，且不会退化为完全均匀分布

- `_LocalAttnHead` / `_GlobalAttnHead` head 数不能整除通道数的边界测试
  - 覆盖 `dim=30`、`num_heads=8`
  - 验证 `_safe_groups` 相关降级逻辑下输出 shape 正确且数值稳定

- `C2fMoA` 多个 `MoABlock` 嵌套时的 aux loss 计数测试
  - 验证 `collect_moa_aux_loss()` 与 `_collect_moa_aux_loss()` 不会重复统计嵌套 block 的辅助损失
  - 防止出现类似 MoE registry 双计数的问题

## 缺陷修复

修改 `ultralytics/nn/modules/moa/moa.py`：

- 修复 `NeckMoAFusion` 的跨尺度对齐逻辑：
  - 将固定 `scale_factor=2` 的上采样改为按 `hi.shape[2:]` 对齐
  - 支持 `15x15 <- 7x7` 这类非严格 2 倍尺度关系

## 测试与覆盖率

测试命令：

```bash
pytest -m tests/test_moa.py
```

结果：

```text
11 passed in 2.92s
```

覆盖率：

```text
Name                                     Stmts   Miss  Cover
----------------------------------------------------------------------
ultralytics/nn/modules/moa/__init__.py       2      0   100%
ultralytics/nn/modules/moa/moa.py          295     35    88%
----------------------------------------------------------------------
TOTAL                                      297     35    88%
```

## 训练验证与实验现象

实验目录：

```text
runs/moa_issue_validation/
```

### 1. MoA 50e 实验

image

`visdrone_moa_50e` 训练稳定：

```text
best mAP50-95 = 0.17170 @ epoch 89
final mAP50-95 = 0.17035
```

### 2. MoE baseline 100e 实验

image

`visdrone_moe_baseline_100e` 训练稳定：

```text
best mAP50-95 = 0.17179 @ epoch 90
final mAP50-95 = 0.17039
```

#### 对比

image

两条基本重合，最终 mAP50-95 是 baseline 0.17039，MoA 0.17035，差 -0.00004，几乎没区别。best 也是 baseline 略高一点：0.17179 @ epoch 90 vs MoA 0.17170 @ epoch 89。

### 异常

image

`visdrone_moa_50e` 中观察到一次明显的数值异常：

```text
epoch 6: val/cls_loss = inf
```

结果：

```text
best mAP50-95 = 0.04452 @ epoch 5
final mAP50-95 = 0.03745
```

该实验在第 6 个 epoch 出现验证分类 loss 爆炸，随后 mAP 长期维持在较低水平，说明该次训练没有稳定收敛。

### CODE REVIEW

**Comment by isLinXu** [comment]
Do not include model weights or logs located under `run/` in your commits; a model-zoo link will be shared later for unified uploads. Please avoid changes to other directories, such as the Ultralytics framework. Only submit content pertaining to `examples/`.

**Comment by isLinXu** [comment]
Do not include model weights or logs located under `run/` in your commits; a model-zoo link will be shared later for unified uploads. Please avoid changes to other directories, such as the Ultralytics framework. Only submit content pertaining to `examples/`.

**Comment by Alanxtl** [comment]
> Do not include model weights or logs located under `run/` in your commits; a model-zoo link will be shared later for unified uploads. Please avoid changes to other directories, such as the Ultralytics framework. Only submit content pertaining to `examples/`.

@isLinXu thank you for your review, I have removed unnecessary `run/`

issue里面要求验证

> NeckMoAFusion 在跨尺度输入尺寸不匹配（如 hi 为 15×15、lo 为 7×7，非严格 2× 下采样）时的前向稳定性与形状保持

之前实现使用固定 `scale_factor=2` 对低分辨率特征进行上采样，因此当 `hi` 为 `15x15`、`lo` 为 `7x7` 这类 odd-size / non-2x feature maps 时，`lo` 只会被上采样到 `14x14`，后续 attention reshape 会因空间 token 数不一致而失败。

根据我的理解这应该不是issue里面的预期行为，我改了 `ultralytics/nn/modules/moa/moa.py`

现在改为根据 `hi.shape[2:]` 使用 `F.interpolate(..., size=hi.shape[2:])` 精确对齐空间尺寸，保证 cross-scale attention 的 `q/k/v` token 数一致，并保持输出 shape 与 `hi` 一致。

**Comment by isLinXu** [comment]
LGTM
